File: origami_master/backend/src/OrigamiMaster.Application/Features/Search/DTOs/SearchResponseDto.cs
using System;
using System.Collections.Generic;

namespace OrigamiMaster.Application.Features.Search.DTOs;

public class SearchResponseDto
{
    public List<SearchUserDto> Users { get; set; } = new();
    public List<SearchPostDto> Posts { get; set; } = new();
    public List<SearchHashtagDto> Hashtags { get; set; } = new();
}

public class SearchHashtagDto
{
    public string Name { get; set; } = string.Empty;
    public int PostCount { get; set; }
}

public class SearchUserDto
{
    public Guid Id { get; set; }
    public string Username { get; set; } = string.Empty;
    public string FullName { get; set; } = string.Empty;
    public string? AvatarUrl { get; set; }
    public int FollowersCount { get; set; }
}

public class SearchPostDto
{
    public Guid Id { get; set; }
    public string Description { get; set; } = string.Empty;
    public string? ImageUrl { get; set; }
    public Guid CreatorId { get; set; }
    public string CreatorName { get; set; } = string.Empty;
    public int LikeCount { get; set; }
}

File: origami_master/frontend/lib/features/home/data/dto/search_response_dto.dart
import 'package:json_annotation/json_annotation.dart';

part 'search_response_dto.g.dart';

@JsonSerializable(createToJson: false)
class SearchResponseDto {
  final List<SearchUserDto> users;
  final List<SearchPostDto> posts;
  final List<SearchHashtagDto> hashtags;

  SearchResponseDto({
    this.users = const [],
    this.posts = const [],
    this.hashtags = const [],
  });

  factory SearchResponseDto.fromJson(Map<String, dynamic> json) =>
      _$SearchResponseDtoFromJson(json);
}

@JsonSerializable(createToJson: false)
class SearchHashtagDto {
  final String name;
  final int postCount;

  SearchHashtagDto({
    required this.name,
    required this.postCount,
  });

  factory SearchHashtagDto.fromJson(Map<String, dynamic> json) =>
      _$SearchHashtagDtoFromJson(json);
}

@JsonSerializable(createToJson: false)
class SearchUserDto {
  final String id;
  final String username;
  final String fullName;
  final String? avatarUrl;
  final int followersCount;

  SearchUserDto({
    required this.id,
    required this.username,
    required this.fullName,
    this.avatarUrl,
    required this.followersCount,
  });

  factory SearchUserDto.fromJson(Map<String, dynamic> json) =>
      _$SearchUserDtoFromJson(json);
}

@JsonSerializable(createToJson: false)
class SearchPostDto {
  final String id;
  final String description;
  final String? imageUrl;
  final String creatorId;
  final String creatorName;
  final int likeCount;

  SearchPostDto({
    required this.id,
    required this.description,
    this.imageUrl,
    required this.creatorId,
    required this.creatorName,
    required this.likeCount,
  });

  factory SearchPostDto.fromJson(Map<String, dynamic> json) =>
      _$SearchPostDtoFromJson(json);
}
